### DISCUSSION TITLE
Questions involving the output data of MultiProgressBar

### DISCUSSION THREAD

**Discussion by joeletho** [JustWhit3/osmanip#26] [General]
Hi @JustWhit3,

I hope you've been well. This is a long one, so I apologize in advance.

I've been working on new functionality of the ```OutputRedirector``` and ```Streambuf``` classes, in which ```osm::cout``` effectively replaces ```std::cout``` and works with ```OutputRedirector``` to redirect the output when being used. Other significant improvements/changes will be included in the next PR.

Until then, I am struggling to understand how the ```MultiProgressBar```'s console and raw output can look so different with the with the same instructions and I'm hoping you can help. Consider this example:
 ```
  osm::redir.begin();
  osm::cout << "These are 3 progress bars printed simultaneously: "
            << "\n\n\n";
  auto bars = MultiProgressBar( prog_int, prog_int_2, prog_float );

  auto job1 = [ &bars, &prog_int ]()
  {
    osm::redir.begin();
    for( int i = prog_int.getMin(); i < prog_int.getMax(); i++ )
    {
      bars.for_one( 0, osm::updater {}, i );
      std::this_thread::sleep_for( std::chrono::milliseconds( 20 ) );
    }
    osm::redir.end();
  };

  //...

  std::thread first_job( job1 );
  //...
  first_job.join();
  //...
  osm::cout << "\n\n\n";
  osm::redir.end();

====================================================
    Console output
====================================================

These are 3 progress bars printed simultaneously:
[#########################] 100% Installing...  
{■■■■■■■■■■■■■■■■■■■■■■■■■} 100% Processing...  
[#########################] 100% Downloading...
```
- Initially, "\n\n\n" is added to the output buffer to make room to draw the progress bars. 
- Then during each loop, the ```for_one()``` method checks the index of the calling bar and assigns a direction in which the cursor should move based on the difference of its index and the previously called bar. 

But, take a look at the raw output (modified for brevity):
```
These are 3 progress bars printed simultaneously: 


[100D[ ... ] 0% Installing... [1B[1B[100D[ ... ] 0% Downloading... [1A[100D{ ... } 0% Processing... [1B[100D[ ...
```
- First we see that that output starts on the third line below the description, instead of immediately below the description as we see in the console output. 
- Then the cursor is to move the left edge of the screen and print ```job1```'s  bar and "Installing..." message.
- Then, the code says to move down two lines before printing the ```job3```'s bar and message. However, the cursor is already at the bottom line and can't move down any further.
- Then up one for ```job2```, down one again for ```job3```, and so on...

Given this output, it's as though the console starts printing on the line just below the description, rather than the bottom row like we see in the ANSI output. Do you know what's going on here and why the console starts printing from the "top" line, rather than the bottom?

Thanks,
Joel

**Comment by JustWhit3** [comment]
Hello @joeletho ,

Sorry if I didn't responded you yet, but this week has been very intense with a lot of work and tomorrow I'll be out of my city for 2 days. I will read and respond your message carefully  on Sunday evening or Monday morning, ok?

Have a nice weekend.

**Reply by joeletho** [reply]
Hi @JustWhit3,

I totally understand. Life catches up with us quickly! I hope you get some time to relax soon.

Take care.

**Comment by JustWhit3** [comment]
Here I am! @joeletho 

I just readed the message and the final output you showed me. I think that the problem is due to the fact that when writing on a txt file you are not able to print ANSI escape sequences on it. And since the bars I used in that example are colored, the ANSI escape sequence of these colors is printed with some strange symbols to the txt file. Same thing holds also for the `\n` character.

So I think that a solution may be to:
- First, reset the progress bar color (with the appropriate setter) before printing the progress bars into a txt file.
- Then, avoid using the `\n` when printing to the same file. This latter is difficult and sincerly I don't know how to solve it know. I have to think about it.

**Reply by joeletho** [reply]
Hi @JustWhit3, I hope you had a good weekend.

Thank you for the response!

Unfortunately, that doesn't answer my question. It's a complicated matter, so I'll try and explain it more succinctly.

For the purposes of the example above, you can ignore the escape sequences and broken symbols and just focus on the "[100D", "[1A", and "[1B" codes, which resemble "left", "up", and "down", respectively. 

I am taking on the (unsurprisingly?) difficult task of "streaming" the output stream buffer into a file and parsing the ANSI escape sequences in place, as they are streamed in. Basically treating the file as the terminal interface. I'm doing this so that the redirected output will look exactly as it looks on the terminal. This will be much more precise than the current implementation.

To summarize, the character is read and added to the file buffer before it is written. If an ANSI escape sequence it is handled according it's command (colors and other styles are ignored). The ANSI escape sequence it not written to the file, only the file buffer is modified as indicated by the ANSI escape sequence.

Ultimately, the confusion I am experiencing is because I'm not understanding how the multibars are printed like they are.
```
 osm::cout << "These are 3 progress bars printed simultaneously: "
           << "\n\n\n";

// Actual output:

These are 3 progress bars printed simultaneously:
[#########################] 100% Installing...  
{■■■■■■■■■■■■■■■■■■■■■■■■■} 100% Processing...  
[#########################] 100% Downloading...

// What the output should look like (if it didn't crash):

These are 3 progress bars printed simultaneously:


[#########################] 100% Installing...  
```

The ```<< "\n\n\n"``` creates the lines needed for the multibars to print and moves up and down the line for each bar. *But* according to the code should see the "Installing" bar printing on the very last line as seen at the bottom of this example, instead of the actual output.
As such, the raw output above confirms that this is the case because there should be a "[2A" command before the "[100D" at the start of the bar, moving the cursor up two lines like the actual output. When designing ```call_one()``` in ```MultiProgressBar``` how did you make this work without moving up before printing the bars?

I can get the actual output by doing the following:
```
 osm::cout << "These are 3 progress bars printed simultaneously: "
           << "\n\n\n";
  osm::cout << osm::feat( crs, "up", 2 );
//...
```
This will move the cursor up to below the message and print the bars.

I know this is confusing --  it's been driving me crazy. My goal in all of this is to make it possible to be able to redirect output just by calling ```.begin()``` and ```.end()``` and I'm very close.

Joel

**Reply by joeletho** [reply]
@JustWhit3,

So, after my latest progress, I can see that you did not have to do anything, it just works. Now, I'm not exactly sure what is going on, but I'm getting similar results to the expected output, which is strange, but OK. Obviously I still have some work to do to further understand what's going on. I'm so close I can taste it! 

Thanks for your help, Gianluca.

Best,

Joel

**Reply by JustWhit3** [reply]
Hi @joeletho,

All's well that ends well! The important thing is that now it is working. However if you have problems understanding something else in the code don't hesitate to ask, I would be glad to help you.

Also if you find something in the functions/classes documentation which is not clear at all for you feel free to change or update it and send the updates in the next pilll request.

Thank you and have a nice day!

**Reply by joeletho** [reply]
Hi @JustWhit3,

Thank you and that's good to know! I may already have a suggestion for ```MultiProgressBar```, but that will be another thread.

Perhaps off topic, but we have an opportunity to choose a paradigm for the redirection and I would like your input.

At the moment, we call ```redir.begin()``` when we want to start redirecting the output and call ```redir.end()``` when we're done. However, we have a few options of how to accomplish this:
- Call ```.begin()``` and ```.end()``` explicitly for each thread to redirect their respective outputs
- Call ```.begin()``` and ```.end()``` and redirect *all* output, regardless of thread
- Have the option to do either and leave it up to the user to decide by setting a flag

I am also thinking that it will be better to make ```OutputRedirector``` global variable. Since ```osm::cout``` works with ```osm::redir``` directly, it only makes sense to do this. Do you have any suggestions for a name? ```osm::redir``` is OK, but has only ever been temporary. ```osm::rout``` could be an option, but I'd like your input.

Thanks!

**Reply by JustWhit3** [reply]
Hi @joeletho ,

Among all your options the second one seems the best one. What do you think? It is in any case less code to write.

For what regard the global variable: maybe `osm::redirout` is speaking itself and seems more intuitive. What do you think about it?

**Reply by joeletho** [reply]
Hi @JustWhit3,

Thanks for the input! I tend to agree with you that ```osm::redirout``` tends to make more sense. And for now we will only need to enable redirection to redirect all output. If we decide we want to enable it on a per thread basis in the future, it's easy to implement.

**Comment by joeletho** [comment]
Hi @JustWhit3,
Just wanna say that the redirection is coming along great
```
First line
Second lineSecond line
Third lineThird line
Fourth line

======================================================
     LOADING BARS                                    
======================================================

This is a normal loading bar: 
[#########################] 

This is a loading bar with message: 
{#########################} processing... 

This is a loading bar with time-remaining info: 
{#########################} [Estimated time left: 0m 0s]


======================================================
     MULTI PROGRESS BARS                                    
======================================================

These are 9 progress bars printed simultaneously: 
[#########################] 100% bar 1 thread... 
{■■■■■■■■■■■■■■■■■■■■■■■■■} 100% bar 2 thread... 
[#########################] 100% bar 4 thread... 
[#########################] 100% bar 4 thread... 
[#########################] 100% bar 7 thread... 
[#########################] 100% bar 6 thread... 
[#########################] 100% bar 7 thread... 
[#########################] 100% bar 8 thread... 
[#########################] 100% bar 9 thread... 

```

I'll make a PR soon.

**Reply by JustWhit3** [reply]
Seems awesome!! Really an excellent work! Thank you very much!
